From 706e46d4e5310820a76bc089def2eb8333695fed Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 12 Jan 2024 16:51:09 +0100
Subject: [PATCH 0423/1695] drivers: stm32_i2c: fix calls to
 stm32_pinctrl_set_secure_cfg()

Move call to stm32_pinctrl_set_secure_cfg() from stm32_i2c_init()
to stm32_i2c_probe(), right before pinctrl configuration is applied
and based on whether to probed I2C bus is secure or not.

Remove useless call to stm32_pinctrl_set_secure_cfg() on
suspend/resume since stm32_gpio driver is in charge of doing that.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 3ae7ed0b8f8e336b116422aac9dd5124585c46ce drivers: stm32_i2c: fix calls to stm32_pinctrl_set_secure_cfg()]
Change-Id: Icb22a2ff6ed8bee6acf4f09dca9a1a392c66725e
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/drivers/stm32_i2c.c | 19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

diff --git a/core/drivers/stm32_i2c.c b/core/drivers/stm32_i2c.c
index 4d2b09870..45d2cd007 100644
--- a/core/drivers/stm32_i2c.c
+++ b/core/drivers/stm32_i2c.c
@@ -296,6 +296,9 @@ struct i2c_request {
 	unsigned int timeout_ms;
 };
 
+/* Non-null reference for compat data */
+static const int secure_i2c;
+
 static vaddr_t get_base(struct i2c_handle_s *hi2c)
 {
 	return io_pa_or_va_secure(&hi2c->base, hi2c->reg_size);
@@ -828,9 +831,6 @@ int stm32_i2c_init(struct i2c_handle_s *hi2c,
 	if (rc)
 		DMSG("I2C analog filter error %d", rc);
 
-	if (IS_ENABLED(CFG_STM32MP13))
-		stm32_pinctrl_set_secure_cfg(hi2c->pinctrl, true);
-
 	clk_disable(hi2c->clock);
 
 	return rc;
@@ -1572,9 +1572,6 @@ void stm32_i2c_resume(struct i2c_handle_s *hi2c)
 
 	restore_cfg(hi2c, &hi2c->sec_cfg);
 
-	if (IS_ENABLED(CFG_STM32MP13))
-		stm32_pinctrl_set_secure_cfg(hi2c->pinctrl, true);
-
 	hi2c->i2c_state = I2C_STATE_READY;
 }
 
@@ -1621,7 +1618,7 @@ static TEE_Result stm32_get_i2c_dev(struct dt_pargs *args, void *data,
 }
 
 static TEE_Result stm32_i2c_probe(const void *fdt, int node,
-				  const void *compat_data __unused)
+				  const void *compat_data)
 {
 	TEE_Result res = TEE_SUCCESS;
 	int subnode = 0;
@@ -1654,6 +1651,9 @@ static TEE_Result stm32_i2c_probe(const void *fdt, int node,
 	init_data.analog_filter = true;
 	init_data.digital_filter_coef = 0;
 
+	if (compat_data == &secure_i2c)
+		stm32_pinctrl_set_secure_cfg(pinctrl_active, true);
+
 	if (pinctrl_apply_state(i2c_handle_p->pinctrl))
 		panic();
 
@@ -1678,9 +1678,10 @@ static TEE_Result stm32_i2c_probe(const void *fdt, int node,
 }
 
 static const struct dt_device_match stm32_i2c_match_table[] = {
-	{ .compatible = "st,stm32mp15-i2c" },
-	{ .compatible = "st,stm32mp13-i2c" },
+	{ .compatible = "st,stm32mp15-i2c", .compat_data = &secure_i2c },
+	{ .compatible = "st,stm32mp13-i2c", .compat_data = &secure_i2c },
 	{ .compatible = "st,stm32mp15-i2c-non-secure" },
+	{ .compatible = "st,stm32mp25-i2c", .compat_data = &secure_i2c },
 	{ }
 };
 
-- 
2.47.2

